tests/testthat/test-generate_data.R

```diff
@@ -381,3 +381,257 @@ test_that("genDummy works.", {
+    rand_id <- sample(nind, 1)
+    rand_state <- sample(nchain, 1)
+    gmnw <- gm_not_wide[id == rand_id & period == rand_state, state]
+    gmw <- gm_wide[rand_id, rand_state + 1, with = FALSE]
+    
+    expect_equal(as.numeric(gmnw), as.numeric(gmw))
+  }
+  
+  for (i in 1:5) {
+    check_equal(gm1, gm1_w)
+    check_equal(gm2, gm2_w)
+  }
+  
+  set.seed(oldSeed)
+  
+})
+
+#genMultiFac ----
+test_that("genMultiFac throws errors.", {
+  oldSeed <- .Random.seed
+  set.seed(98765)
+  
+  # check nFactors are integers
+  expect_error(genMultiFac(1.4, each = 4), class = "simstudy::wrongType")
+  
+  # check length nFactors greater than 2
+  expect_error(genMultiFac(1, each = 4), class = "simstudy::greaterThan")
+  
+  # check number of levels matches factors
+  expect_error(genMultiFac(3, levels = c(2, 3)), class = "simstudy::notEqual")
+  
+  # check coding == 'effect' or 'dummy'
+  expect_error(genMultiFac(2, each = 3, coding = "trtAssign"), class = "simstudy::codingVal")
+  
+  set.seed(oldSeed)
+})
+
+test_that("genMultiFac works.", {
+  oldSeed <- .Random.seed
+  set.seed(87654)
+  
+  # coding == dummy, levels == 2
+  nFac <- sample(2:5, size = 1)
+  nEach <- sample(2:5, size = 1)
+  g1 <- genMultiFac(nFac, each = nEach)
+  g1
+  for(i in sample(2:length(g1))) {
+    expect_equal(sum(g1[, i, with = FALSE]), nrow(g1) / 2)
+  }
+  
+  # this is from stackoverflow, is that an issue with publishing?
+  pascalTriangle <- function(h) {
+    lapply(0:h, function(i) choose(i, 0:i))
+  }
+  
+  pt <- pascalTriangle(nFac)
+  
+  poss_row_sums1 <- NULL
+  
+  for(j in 0:nFac) {
+    pt_num <- pt[[nFac + 1]][j + 1]
+    for(k in 1:pt_num) {
+      for(i in 1:nEach) {
+        poss_row_sums1 <- c(poss_row_sums1, j)
+      }
+    }
+  }
+  
+  actual_row_sums1 <- NULL
+  for(i in 1:nrow(g1)) {
+    actual_row_sums1 <- c(actual_row_sums1, sum(g1[i, 2:length(g1)]))
+  }
+  
+  poss_row_sums1 <- sort(unlist(poss_row_sums1))
+  actual_row_sums1 <- sort(unlist(actual_row_sums1))
+  
+  expect_true(all(poss_row_sums1 == actual_row_sums1))
+  expect_true(all(g1[, 2:nFac] == 0 | g1[, 2:nFac] == 1))
+  
+  
+  ## coding == effect, levels == 2
+  nFac <- sample(2:5, size = 1)
+  nEach <- sample(2:5, size = 1)
+  g2 <- genMultiFac(nFac, each = nEach, coding = "effect")
+  for(i in sample(2:length(g2))) {
+    expect_equal(sum(g2[, i, with = FALSE]), 0)
+  }
+  expect_true(all(g2[, 2:nFac] == 1 | g2[, 2:nFac] == -1))
+  
+  
+  ## levels == other, len(levels) == 1
+  nFac <- sample(2:5, size = 1)
+  nEach <- sample(2:5, size = 1)
+  nLev <- sample(2:5, size = 1)
+  g3 <- genMultiFac(nFac, each = nEach, levels = nLev)
+  
+  sumTable <- table(rowSums(g3[, 2:length(g3)]))
+  # checks all unique row sums are accounted for
+  for(i in nFac:(nFac * nLev)) {
+    ind <- as.character(i)
+    expect_true(!is.na(sumTable[ind]))
+  }
+  
+  # checks each column has appropriate number of each value in it (1s, 2s, 3s, etc.)
+  for(i in 2:nFac + 1) {
```

Comment:
This should be `for(i in 2:(nFac + 1)) {`

I still have concerns with this approach, because it seems hard to 'prove' that it would work in all cases. As I said earlier, it seems clear that if the data generation works properly, this will pass. It isn't so obvious that there is no be failed data generation process that could pass. If you can prove - or at least argue very strongly - that _**only**_ a correct randomization will pass - then I will completely trust this. If that proof or argument is not so obvious, then taking another approach may be easier.

One quick thought I had was to convert the sequence of numbers in each case to a single string of numbers, and then use some sort of base-n conversion to convert that sequence to base 10. So, for example, if we have

```
1 1 1
1 1 2
1 1 3
.
.
.
3 3 2
3 3 3
```

Let's say we convert that (by subtracting 1 from everything) to

```
0 0 0
0 0 1
0 0 2
.
.
.
2 2 1
2 2 2
```

Now we have a sequence of numbers in base-3, `000, 001, ... 222`, which can be converted to '0, 1, .. 26`, and if we add 1 to that, we have 1 through 27, and each is unique - we need to generate the data correctly to get that. Maybe this is not a proof, but it is a compelling argument, and I have no doubt that it would pick up errors.

Maybe you could implement something along these lines?

Reply:
I see, I had convinced myself that this was pretty airtight but looking at it more I can see a way where a false generation would pass. What was advantageous about the Pascal's triangle idea is it would just look for the unique sums, but for the number of each sums, which is necessary to prove that the table is generated properly. 

I like the idea you suggested and I'll try writing that in

Reply:
How does what I did look? I believe it achieves the same idea as what you were saying.

Reply:
I think that could work - even without converting to base 10, as long as you check to make sure there are no categories outside the expected range.

You can replace the loop (which are quite slow in R) with this single `lapply` command. I know it is a little ugly, but it is probably faster:

```
rowStrings <- unlist(lapply(split(g3[,-1], seq(nrow(g3))), function(x) paste0(x, collapse="")))
```

So the whole check reduces to this:

```
nFac <- sample(2:5, size = 1)
nEach <- sample(2:5, size = 1)
nLev <- sample(2:5, size = 1)
g3 <- genMultiFac(nFac, each = nEach, levels = nLev)

rowStrings <- unlist(lapply(split(g3[,-1], seq(nrow(g3))), function(x) paste0(x, collapse="")))
expect_true(length(rowStrings) == (length(unique(rowStrings)) * nEach))
```

Of course, you still need to check that all values are in the correct range ...

Reply:
I believe I hit both of those

Reply:
I haven't looked at the latest code, but can you clean up, and make sure it includes only the final tests. I guess that means, as cool as the Pascal triangle idea was, you should probably take those out (if you haven't already).

Reply:
Yes for sure